File: 15.三数之和.cs
/*
 * @lc app=leetcode.cn id=15 lang=csharp
 *
 * [15] 三数之和
 */

// @lc code=start
public partial class Solution
{
    public IList<IList<int>> ThreeSum(int[] nums)
    {
        IList<IList<int>> result = new List<IList<int>>();

        HashSet<int> num1s = new HashSet<int>();
        for (int i = 0; i < nums.Length - 2; i++)
        {
            int num1 = nums[i];
            if (num1s.Contains(num1))
            {
                continue;
            }
            int outerTarget = -num1;
            Dictionary<int, bool> num2s = new Dictionary<int, bool>();
            for (int j = i + 1; j < nums.Length; j++)
            {
                int num2 = nums[j];
                if (num1s.Contains(num2))
                {
                    continue;
                }
                if (num2s.TryGetValue(num2, out bool success) && success)
                {
                    continue;
                }
                int num3 = outerTarget - num2;
                if (num1s.Contains(num3))
                {
                    continue;
                }
                if (num2s.ContainsKey(num3))
                {
                    result.Add(new List<int>() { num1, num2, num3 });
                    num2s[num2] = num2s[num3] = true;
                }
                else
                {
                    num2s[num2] = false;
                }
            }
            num1s.Add(num1);
        }

        return result;
    }
}
// @lc code=end

File: src/15.三数之和.rs
/*
 * @lc app=leetcode.cn id=15 lang=rust
 *
 * [15] 三数之和
 */

// @lc code=start
impl Solution {
    pub fn three_sum(mut nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut result: Vec<Vec<i32>> = Vec::new();
        nums.sort();

        for i in 0..nums.len() {
            let num1 = nums[i];
            if i != 0 {
                if num1 == nums[i - 1] {
                    continue;
                }
            }
            let target = -num1;
            let mut left = i + 1;
            let mut right = nums.len() - 1;
            while left < right {
                let num2 = nums[left];
                let num3 = nums[right];
                if (num2 + num3) <= target {
                    if (num2 + num3) == target {
                        result.push(vec![num1, num2, num3]);
                        right = nums.len() - 1;
                    }
                    loop {
                        left += 1;
                        if left >= right || nums[left] > num2 {
                            break;
                        }
                    }
                } else {
                    loop {
                        right -= 1;
                        if left >= right || nums[right] < num3 {
                            break;
                        }
                    }
                }
            }
        }

        return result;
    }
}
// @lc code=end
